### PR TITLE
Fixes #8316: Removing a host can create a huge transaction

### PR DESCRIPTION
**Pull request by dLobatog**
Removing a host with some reports (say 200), and each of them with several log lines (10000) generates 2.000.000 DELETE sql statements in the same transaction.
More often than not, ActiveRecord won't be able to make such a big transaction, in fact it will likely eat all memory available (I tried on a machine 16GB RAM machine) and Foreman will stop working in that host. In the best scenario the DB will end up with a non-committed giant transaction.

I suggest deleting hosts and reports manually on host destroy to avoid this.

### CODE REVIEW

File: app/models/host/base.rb [hunk 1]
```diff
@@ -240,5 +241,12 @@ def interface_class(name)
           Nic::Managed
       end
     end
+
+    # For performance reasons reports and logs are deleted in batch,
+    # see http://projects.theforeman.org/issues/8316 for details
+    def remove_reports
+      reports.each { |report| Log.delete_all("report_id = #{report.id}") }
```
Comment (tbrisker) [inline app/models/host/base.rb]: Wouldn't it be better to use `find_each` instead of `each` here, in case the host has a very large number of reports?

Comment (dLobatog) [inline app/models/host/base.rb]: That sounds good, if we're going to do it efficiently let's do it as efficiently as possible.

Comment (ohadlevy) [inline app/models/host/base.rb]: then use an inner query instead? :)

**Comment by dLobatog** [comment]
Updated, I use `reports.pluck(:id)` since `SELECT reports.id` is faster than `SELECT reports.*`. Also the reports association is defined for managed hosts only so I defined the callback there.

@ohadlevy Inner query?

File: app/models/host/managed.rb [hunk 2]
```diff
@@ -963,4 +964,11 @@ def update_hostgroups_puppetclasses
     Hostgroup.find(hostgroup_id).update_puppetclasses_total_hosts     if hostgroup_id.present?
   end
 
+  # For performance reasons logs and reports are deleted in batch
+  # see http://projects.theforeman.org/issues/8316 for details
+  def remove_reports
+    return if reports.empty?
+    Log.delete_all("report_id IN (#{reports.pluck(:id).join(',')})")
```
Comment (lzap) [inline app/models/host/managed.rb:930]: But won't this hit limit of numbers that can reside in the "in" statement? If we can, we should really use nested queries (I have no clue how to do this in AR tho):

``` sql
DELETE FROM logs
WHERE report_id IN 
    (SELECT id FROM reports
     WHERE log_id = ?)
```

**Comment by lzap** [comment]
Quick google reveals that it should be possible:

``` ruby
# 1. Build you subquery with AREL.
subquery = Account.where(...).select(:id)
# 2. Use the AREL object in your query by converting it into a SQL string
query = User.where("users.account_id IN (#{subquery.to_sql})")
```

File: app/models/host/managed.rb (hunk 2 above)
Comment (ohadlevy) [inline app/models/host/managed.rb:930]: yes, this is what i had in mind too, in rails its simple:

``` ruby
Log.where(:report_id => Report.recent)
```

will generate

``` sql
SELECT "logs".* FROM "logs" WHERE "logs"."report_id" IN (SELECT "reports"."id" FROM "reports" WHERE (reported_at > '2014-11-17 11:36:53.150355') ORDER BY reported_at)
```

Also, there is a host expire reports method (when clicking on build it removes old reports & facts) maybe that should be unified.

Comment (dLobatog) [inline app/models/host/managed.rb:930]: Shouldn't that be:

```
DELETE FROM logs
WHERE report_id IN 
    (SELECT id FROM reports
     WHERE host_id = ?)
```

?

Comment (ohadlevy) [inline app/models/host/managed.rb:930]: it was just an example for inner query, of course replace with the relevant code :)
you can just add .delete_all at the end afair.

you can probably refactor https://github.com/theforeman/foreman/blob/develop/app/models/report.rb#L123-L126 to include a single host as well.

**Comment by dLobatog** [comment]
@ohadlevy @lzap 

I've just tested this code:

``` ruby
    inner_query = Report.where(:host_id => id).select(:id)
    Log.delete_all("report_id IN (#{inner_query.to_sql})")
    Report.delete_all("host_id = #{id}")
```

and it's unacceptably slow (9258 logs, 198 reports):

``` sql
  SQL (79151.9ms)  DELETE FROM `logs` WHERE (report_id IN (SELECT id FROM `reports` WHERE `reports`.`host_id` = 7))
  SQL (140.9ms)  DELETE FROM `reports` WHERE (host_id = 7)
```

By comparison, take a look at the same operation with the current code in the same host:

``` sql
  SQL (767.1ms)  DELETE FROM `logs` WHERE (report_id IN (20449266,20453254,20457442,20461391,20465340,20469467,20473522,20477353,20481550,20485664,20489534,20493583,20497766,20501637,20505605,20509846,20513797,20518028,20522448,20526937,20531221,20535741,20539817,20543983,20547866,20551895,20556094,20560062,20563954,20568167,20572252,20576129,20580278,20584415,20588350,20592516,20596829,20600810,20604942,20609194,20613319,20617692,20622190,20626849,20631256,20635796,20639965,20644337,20648252,20652275,20656563,20660636,20664646,20668911,20672943,20676828,20681009,20685121,20689028,20693125,20697360,20701308,20705283,20709540,20713522,20717842,20722245,20726750,20731111,20735931,20739923,20744198,20748192,20752089,20756275,20760391,20764269,20768441,20772568,20776482,20780554,20784831,20788689,20792717,20796974,20800990,20804916,20809105,20813183,20817520,20821898,20826489,20830884,20835579,20839548,20843754,20847779,20851728,20855845,20860026,20863861,20867980,20872208,20876042,20880116,20884368,20888260,20892268,20896469,20900487,20904449,20908615,20912739,20917126,20921461,20926081,20930428,20933912,20938029,20942267,20946104,20950149,20954448,20958424,20962463,20966677,20970804,20974759,20978976,20983308,20987298,20991503,20995820,20999707,21003847,21008157,21012192,21016607,21021066,21025696,21030155,21034941,21039093,21043435,21047382,21051451,21055738,21059849,21063933,21068288,21072521,21076446,21080787,21085220,21089172,21093343,21097719,21101768,21105881,21110237,21114338,21118942,21123307,21127972,21132606,21137462,21141538,21145950,21150035,21154067,21158410,21162706,21166711,21171045,21175358,21179409,21183657,21188078,21192217,21196486))
  SQL (287.7ms)  DELETE FROM `reports` WHERE (host_id = 7)
```

There are a few solutions to avoid a large IN-clause like the one above, such as [using intermediate tables](http://explainextended.com/2009/08/18/passing-parameters-in-mysql-in-list-vs-temporary-table/), but I think it's overkill in this case. The default maximum number of ids in an IN clause is [defined in the manual](http://dev.mysql.com/doc/refman/5.1/en/comparison-operators.html#function_in) and it's 1048576 (1MB). 

Keep in mind the number of IDs we're talking about here is the number of reports per host. Logs could reach that number in a large installation that keeps reports for a very long time. Just to illustrate it, a host getting reports every 20 minutes would require nearly 10 years until the limit is hit.

```
3 (reports per hour) * 24 * 30 * 12 = **25920 reports**
```

Any other ideas? I think the current code might not be the safest but works very fast, removing 1000 hosts with this number of reports would take 16s, which is unlikely to timeout.

**Comment by lzap** [comment]
Hmmm were you testing on PostgreSQL or SQLite?

**Comment by lzap** [comment]
Roughly reviewing our migrations, I don't see index on `reports.host_id`, only combined ones. Can you try this:

``` ruby
    inner_query = Report.where(:host_id => id, :type => Host::Managed).select(:id)
    Log.delete_all("report_id IN (#{inner_query.to_sql})")
    Report.delete_all("host_id = #{id}")
```

**Comment by lzap** [comment]
Disregard, I see the index now:

  add_index "reports", ["host_id"], :name => "index_reports_on_host_id"

Can you just verify it's present in our database? This is strange behavior, this must work as fast as possible.

**Comment by lzap** [comment]
I think I've lost the thread a bit. Why can't we use something like:

``` sql
DELETE FROM reports WHERE host_id = 42;
```

Which is, I believe, generated by:

``` ruby
Report.where(:host_id => 42).destroy_all
```

I can't remember if there is need to join a table, but if this is the case, then most SQL servers can do something like:

``` sql
DELETE FROM reports INNER JOIN hosts ON (reports.host_id = hosts.id) WHERE hosts.name = 'xyz';
```

Not sure how to write this in Ruby, maybe it will work:

Report.joins(:hosts).where(:host => self).destroy_all

**Comment by dLobatog** [comment]
@lzap I think you're confused between `destroy` and `delete`. The SQL query you mention in that comment would be generated by `delete_all`. Using `destroy_all` instead is going to trigger `.destroy` for associated objects, invoking callbacks, like :dependent and so forth. That's going to generate the same query as Host.find(42).destroy will generate, which looks like this (i've just ran it):

```
SELECT `reports`.* FROM `reports` WHERE `reports`.`host_id` = 19221
  Log Load (1.4ms)  SELECT `logs`.* FROM `logs` WHERE `logs`.`report_id` = 21187922
  SQL (0.3ms)  DELETE FROM `logs` WHERE `logs`.`id` = 728379118
  ........................many deletes after..................................
  SQL (0.1ms)  DELETE FROM `logs` WHERE `logs`.`id` = 728379219
  SQL (0.4ms)  DELETE FROM `reports` WHERE `reports`.`id` = 21187922
  ......................repeat this for every report.............................
```

As I explain in the first post or the Redmine issue, this will likely clog the database quickly, especially if you're trying to delete many hosts at the same time. The difference in this pull request is that I'm using `delete_all` which generates that SQL code you wrote above, a simple DELETE query, for logs and for reports.

**Comment by lzap** [comment]
Yeah, I am always confused with destroy vs delete. But that does not change my idea. Why can't we simply do this:

```
Report.where(:host_id => 42).delete_all
```

I mean something to generate the following SQL:

```
DELETE FROM reports WHERE host_id = 42;
```

We know the host id already and reports has host_id column. We do not need to iterate or fetch the reports.

**Comment by lzap** [comment]
And how about the DELETE with INNER JOIN? Will that work?

**Comment by dLobatog** [comment]
@lzap I'm doing exactly that:

``` ruby
Report.delete_all("host_id = #{id}")
```

is an equivalent of what you wrote above (or at least it generates the same query). I don't iterate over all reports of anything, that's what I'm trying to avoid. I do however fetch the ids of all reports to remove all Logs whose report_id matches that, because `delete_all` won't delete the logs, only the reports.

**Comment by lzap** [comment]
On the IRC you mentioned you are testing with MySQL. I am doing some PostgreSQL analysis and subselect does not look bad at all. I've created 5000 reports with one log each and it takes a fraction of second to delete all (4999) the records:

```
for X in {1..5000}; do echo "INSERT INTO reports (host_id, reported_at, created_at, updated_at, status) VALUES (1, current_timestamp, current_timestamp, current_timestamp, 100663361);" | psql foreman &>/dev/null ; done

bash-4.1$ for X in {1..5000}; do echo "INSERT INTO logs (source_id, message_id, report_id, level_id ,    created_at, updated_at) VALUES (121,227,$X,2,current_timestamp,current_timestamp);" | psql foreman &>/dev/null ; done

bash-4.1$ echo "EXPLAIN ANALYZE DELETE FROM logs WHERE (report_id IN (SELECT id FROM reports WHERE reports.host_id = 1));" | time psql foreman
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Hash Semi Join  (cost=167.40..352.15 rows=4936 width=6) (actual time=17.584..41.242 rows=4999 loops=1)
   Hash Cond: (logs.report_id = reports.id)
   ->  Seq Scan on logs  (cost=0.00..92.00 rows=5000 width=10) (actual time=0.028..8.209 rows=5000 loops=1)
   ->  Hash  (cost=105.70..105.70 rows=4936 width=4) (actual time=17.510..17.510 rows=5025 loops=1)
         ->  Seq Scan on reports  (cost=0.00..105.70 rows=4936 width=4) (actual time=0.025..8.851 rows=5025 loops=1)
               Filter: (host_id = 1)
 Total runtime: 52.865 ms
(7 rows)
```

I guess MySQL needs to be optimized a bit (not to use InnoDB for logs/reports perhaps - MyISAM could be better). But we should _not_ optimize Foreman Core for MySQL. Our target platform _for optimalizations_ is PostgreSQL IMHO.

**Comment by lzap** [comment]
Re-tested with 100 reports with 100 logs each:

```
bash-4.1$ echo "EXPLAIN ANALYZE DELETE FROM logs WHERE (report_id IN (SELECT id FROM reports WHERE reports.host_id = 1));" | time psql foreman
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Hash Semi Join  (cost=167.40..539.04 rows=10005 width=6) (actual time=17.269..64.515 rows=9900 loops=1)
   Hash Cond: (logs.report_id = reports.id)
   ->  Seq Scan on logs  (cost=0.00..184.05 rows=10005 width=10) (actual time=0.027..16.288 rows=10002 loops=1)
   ->  Hash  (cost=105.70..105.70 rows=4936 width=4) (actual time=17.191..17.191 rows=5028 loops=1)
         ->  Seq Scan on reports  (cost=0.00..105.70 rows=4936 width=4) (actual time=0.024..8.695 rows=5028 loops=1)
               Filter: (host_id = 1)
 Total runtime: 88.231 ms
(7 rows)
```

**Comment by dLobatog** [comment]
The results I get with MariaDB 10.0.15 when I run it through the Rails console from what I get running the query directly:

Running the query on my own:

```
EXPLAIN DELETE FROM `logs` WHERE (report_id IN (SELECT id FROM `reports` WHERE `reports`.`host_id` = 8))
    -> ;
+------+--------------------+---------+-----------------+---------------------------------------------------------------------------+---------+---------+------+----------+-------------+
| id   | select_type        | table   | type            | possible_keys                                                             | key     | key_len | ref  | rows     | Extra       |
+------+--------------------+---------+-----------------+---------------------------------------------------------------------------+---------+---------+------+----------+-------------+
|    1 | PRIMARY            | logs    | ALL             | NULL                                                                      | NULL    | NULL    | NULL | 36663383 | Using where |
|    2 | DEPENDENT SUBQUERY | reports | unique_subquery | PRIMARY,index_reports_on_reported_at_and_host_id,index_reports_on_host_id | PRIMARY | 4       | func |        1 | Using where |
+------+--------------------+---------+-----------------+---------------------------------------------------------------------------+---------+---------+------+----------+-------------+
2 rows in set (0.01 sec)

MariaDB [foreman]> DELETE FROM `logs` WHERE (report_id IN (SELECT id FROM `reports` WHERE `reports`.`host_id` = 8));
Query OK, 9011 rows affected (1 min 17.10 sec)
```

Running the query on the Rails console:

```
2.0.0-p481 :011 >     Log.delete_all("report_id IN (#{inner_query.to_sql})")
  SQL (84384.6ms)  DELETE FROM `logs` WHERE (report_id IN (SELECT id FROM `reports` WHERE `reports`.`host_id` = 8))
```

**Comment by domcleal** [comment]
I'd like to just interject that very similar code is over in https://github.com/theforeman/foreman/blob/1.7.0/app/models/report.rb#L123, and we have a bug filed to say that it's running very slowly: http://projects.theforeman.org/issues/8565

While we're waiting on precise debug details for that bug, it might be worth waiting to see which line is at fault in case we introduce a second problem in this change.

**Comment by lzap** [comment]
@eLobato looks like mis-optimalization bug in MySQL which was accepted and will likely be fixed in the future. In the meantime, can you try the workaround described there: http://bugs.mysql.com/bug.php?id=35794

``` sql
delete from carts left join line_items on(carts.id=line_items.cart_id) where line_items.cart_id is null;
```

If it works, I am not sure if this is worth the effort fixing this. I'd rather proceed with documenting this behavior.

**Comment by domcleal** [comment]
Ok, don't worry about the other bug (https://github.com/theforeman/foreman/pull/1907#issuecomment-66128318), seems it was the following lines that were slow and not the report delete line.

**Comment by ohadlevy** [comment]
@eLobato if i follow the conversation correctly, inner select is the way to go?

File: app/models/host/managed.rb [hunk 3]
```diff
@@ -963,4 +964,11 @@ def update_hostgroups_puppetclasses
     Hostgroup.find(hostgroup_id).update_puppetclasses_total_hosts     if hostgroup_id.present?
   end
 
+  # For performance reasons logs and reports are deleted in batch
+  # see http://projects.theforeman.org/issues/8316 for details
+  def remove_reports
```
Comment (ohadlevy) [inline app/models/host/managed.rb:928]: when changing to build mode, reports and facts get flushed too, can we reuse the same method?

File: app/models/host/managed.rb (hunk 2 above)
Comment (unorthodoxgeek) [inline app/models/host/managed.rb:930]: should be this:

``` ruby
Log.where(report_id:reports).delete_all
reports.delete_all
```

File: app/models/host/managed.rb (hunk 3 above)
Comment (dLobatog) [inline app/models/host/managed.rb:928]: I think it can be reused yes.

**Comment by dLobatog** [comment]
@ohadlevy I'm not sure, I'd say no until MySQL/MariaDB perform well with inner selects, otherwise we're just solving the problem for PostgreSQL. The current fix works for both.

**Comment by lzap** [comment]
If we are going to accept the "pluck-hack" which I am not totally against (it just does not look correct when PostgreSQL can handle this correctly and MySQL eventually too), put all the ARES statements within the `remove_reports` into one SQL transaction please. Let's have all or nothing.

**Comment by dLobatog** [comment]
Rebased @lzap @ohadlevy

File: app/models/host/managed.rb (hunk 3 above)
Comment (ohadlevy) [inline app/models/host/managed.rb:928]: hint: lets use it then? :)

**Comment by dLobatog** [comment]
Rebased again @lzap @ohadlevy

**Comment by lzap** [comment]
Works fine. Not ideal, but better than nothing. Thanks for patience with us :-)

```
Host.all.each {|h| h.send :remove_reports }
  Host::Managed Load (0.7ms)  SELECT "hosts".* FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed')
   (0.1ms)  SELECT COUNT(*) FROM "reports" WHERE "reports"."host_id" = 46
   (0.2ms)  SELECT "reports"."id" FROM "reports" WHERE "reports"."host_id" = 46
  SQL (21.3ms)  DELETE FROM "logs" WHERE (report_id IN (3,4,5,6,7))
  SQL (40.1ms)  DELETE FROM "reports" WHERE (host_id = 46)
```

Merged as 3e75071.

**Pull request closed without merging**